base/stream.jl

```diff
@@ -484,11 +484,39 @@ function displaysize(io::TTY)
     w > 0 || (w = default_size[2])
     return h, w
 end
-
-in(key_value::Pair{Symbol,Bool}, ::TTY) = key_value.first === :color && key_value.second === have_color
-haskey(::TTY, key::Symbol) = key === :color
-getindex(::TTY, key::Symbol) = key === :color ? have_color : throw(KeyError(key))
-get(::TTY, key::Symbol, default) = key === :color ? have_color : default
+function ttyhascolor()
+    term_type = get(ENV, "TERM","")
+    startswith(term_type, "xterm") && return true
+    try
+        @static if Sys.KERNEL === :FreeBSD
+            return success(`tput AF 0`)
+        else
+            return success(`tput setaf 0`)
+        end
+    catch
+        return false
+    end
+end
+function in(key_value::Pair{Symbol,Bool}, ::TTY)
+    global have_color
+    have_color == nothing && (have_color = ttyhascolor())
+    return key_value.first === :color && key_value.second === have_color
+end
+function haskey(::TTY, key::Symbol)
+    global have_color
+    have_color == nothing && (have_color = ttyhascolor())
```

Comment:
Don't repeat this code. You can move it to its own function, then replace uses of `have_color` with `get_have_color()`:
```
function get_have_color()
    global have_color
    have_color === nothing && (have_color = ttyhascolor())
    return have_color::Bool
end
```

Reply:
Yes, this looks quite better , thanks!